src/nodes/CompositeNode.jl

```diff
@@ -0,0 +1,249 @@
+@doc doc"""
+```julia
+CompositeNode(CurrentNodes)
+```
+
+A composite node consisting of current and power sources provided, as well as up to one
+optional voltage controling node. This assumes that the
+current nodes are implemented as
+
+```julia
+du = i - I_node
+```
+
+and the power nodes are implemented with
+
+```julia
+du = u*conj(i) - P_node
+```
+
+*Important!!* The individual nodes are given the current that is arriving from
+the grid into the bus, _not_ the current they themself inject! Check the definitions
+of the nodes going into the CompositeType to make sure they have the right dynamics.
+
+*Experimental!!* This node has not been sufficiently tested for correctness yet.
+
+"""
+
+struct _IPU end
+struct _IU end
+struct _PU end
+struct _IP end
+struct _I end
+struct _P end
+
+
+CompositionTypes = Dict(["IPU" => _IPU,
+    "IU" => _IU,
+    "PU" => _PU,
+    "I" => _I,
+    "P" => _P,
+    "IP" => _IP])
+
+@__doc__ mutable struct CompositeNode{T}
+    CurrentNodes # constant current nodes
+    VoltageNode # constant voltage nodes
+    PowerNodes # constant power nodes
+    internaldims::Array{Int, 1}
+    total_dim::Int
+    idxs::Array{Array{Int,1},1}
+    symbols::Array{Symbol, 1}
+    function CompositeNode(;CurrentNodes=nothing, VoltageNode=nothing, PowerNodes=nothing)# PowerNodes, VoltageNode)
+
+        @assert any([CurrentNodes != nothing, PowerNodes != nothing])
+
+        @assert !(typeof(VoltageNode) <: AbstractArray)
+
+        symbols = [:u_r, :u_i]
+        internaldims = Array{Int,1}()
+        mix = ""
+
+        # The key implementation detail is that we stack the internal variables
+        # of all nodes in the order I, P and U. The array of indices makes sure
+        # that these are combined correctly with the first two indices that represent
+        # the voltage or the current constraints.
+
+        if CurrentNodes != nothing
+            mix *= "I"
+            for current_node in CurrentNodes
+                append!(symbols, symbolsof(current_node)[3:end])
```

Comment:
what if a current node has no additional internal variables? like the CSIMinimal?

Reply:
Then this append! does nothing.
```julia
a = [1,2]
b = [6,7]
append!(a, b[3:2])
append!(a, b[3:end])
a == [1,2] # true
```